build_resolvers/lib/src/build_asset_uri_resolver.dart

```diff
@@ -11,70 +12,81 @@ import 'package:analyzer/file_system/memory_file_system.dart';
 import 'package:build/build.dart' show AssetId, BuildStep;
+import 'package:crypto/crypto.dart';
+import 'package:graphs/graphs.dart';
 import 'package:path/path.dart' as p;
 
 const _ignoredSchemes = ['dart', 'dart-ext'];
 
 class BuildAssetUriResolver extends UriResolver {
+  /// A cache of the directives for each Dart library.
+  ///
+  /// This stored across builds and is only invalidated if we read a file and
+  /// see that it's content is different from what it was last time it was read.
   final _cachedAssetDependencies = <AssetId, Set<AssetId>>{};
-  final _cachedAssetContents = <AssetId, String>{};
+
+  /// A cache of the digest for each Dart asset.
+  ///
+  /// This is stored across builds and used to invalidate the values in
+  /// [_cachedAssetDependencies] only when the actual content of the library
+  /// changed.
+  final _cachedAssetDigests = <AssetId, Digest>{};
+
   final resourceProvider = MemoryResourceProvider(context: p.posix);
 
-  /// The assets which are known to be readable at some point during the build.
+  /// The assets which are known to be readable at some point during the current
+  /// build.
   ///
   /// When actions can run out of order an asset can move from being readable
   /// (in the later phase) to being unreadable (in the earlier phase which ran
   /// later). If this happens we don't want to hide the asset from the analyzer.
-  final seenAssets = Set<AssetId>();
-
-  Future<void> performResolve(
-      BuildStep buildStep, List<AssetId> entryPoints, AnalysisDriver driver) {
-    // Basic approach is to start at the first file, update it's contents
-    // and see if it changed, then walk all files accessed by it.
-    var visited = Set<AssetId>();
-    var visiting = _FutureGroup();
-
-    void processAsset(AssetId assetId) {
-      visited.add(assetId);
-
-      visiting.add(buildStep.readAsString(assetId).then((contents) {
-        if (_cachedAssetContents[assetId] != contents) {
-          if (_cachedAssetContents.containsKey(assetId)) {
-            var path = assetPath(assetId);
-            resourceProvider.updateFile(path, contents);
-            driver.changeFile(path);
-          } else {
-            resourceProvider.newFile(assetPath(assetId), contents);
-          }
-          _cachedAssetContents[assetId] = contents;
-          var unit = parseDirectives(contents, suppressErrors: true);
-          var dependencies = unit.directives
-              .whereType<UriBasedDirective>()
-              .where((directive) {
-                var uri = Uri.parse(directive.uri.stringValue);
-                return !_ignoredSchemes.any(uri.isScheme);
-              })
-              .map((d) => AssetId.resolve(d.uri.stringValue, from: assetId))
-              .where((id) => id != null)
-              .toSet();
-          _cachedAssetDependencies[assetId] = dependencies;
+  final seenAssets = HashSet<AssetId>();
+
+  Future<void> performResolve(BuildStep buildStep, List<AssetId> entryPoints,
```

Comment:
Can we get a doc comment on this function?

Reply:
Done